crates/autopilot/src/domain/competition/winner_selection/combinatorial.rs

```diff
@@ -0,0 +1,352 @@
+                },
+                score,
+            );
+            p.set_computed_score(total_score);
+            true
+        }
+        Err(err) => {
+            tracing::warn!(
+                driver = p.driver().name,
+                ?err,
+                solution = ?p.solution(),
+                "discarding solution where scores could not be computed"
+            );
+            false
+        }
+    });
+
+    scores
+}
+
+/// Returns the total scores for each directed token pair of the solution.
+/// E.g. if a solution contains 3 orders like:
+///     sell A for B with a score of 10
+///     sell A for B with a score of 5
+///     sell B for C with a score of 5
+/// it will return a map like:
+///     (A, B) => 15
+///     (B, C) => 5
+fn score_by_token_pair(solution: &Solution, auction: &Auction) -> Result<ScoreByDirection> {
+    let mut scores = HashMap::default();
+    for (uid, trade) in solution.orders() {
+        if !auction.contributes_to_score(uid) {
+            continue;
+        }
+
+        let uniform_sell_price = solution
+            .prices()
+            .get(&trade.sell.token)
+            .context("no uniform clearing price for sell token")?;
+        let uniform_buy_price = solution
+            .prices()
+            .get(&trade.buy.token)
+            .context("no uniform clearing price for buy token")?;
+
+        let trade = math::Trade {
+            uid: *uid,
+            sell: trade.sell,
+            buy: trade.buy,
+            side: trade.side,
+            executed: match trade.side {
+                order::Side::Buy => TargetAmount(trade.executed_buy.into()),
+                order::Side::Sell => TargetAmount(trade.executed_sell.into()),
+            },
+            prices: transaction::Prices {
+                // clearing prices are denominated in the same underlying
+                // unit so we assign sell to sell and buy to buy
+                uniform: ClearingPrices {
+                    sell: uniform_sell_price.get().into(),
+                    buy: uniform_buy_price.get().into(),
+                },
+                // for custom clearing prices we only need to know how
+                // much the traded tokens are worth relative to each
+                // other so we can simply use the swapped executed
+                // amounts here
+                custom: ClearingPrices {
+                    sell: trade.executed_buy.into(),
+                    buy: trade.executed_sell.into(),
+                },
+            },
+        };
+        let score = trade
+            .score(&auction.fee_policies, auction.native_prices)
+            .context("failed to compute score")?;
+
+        let token_pair = DirectedTokenPair {
+            sell: trade.sell.token,
+            buy: trade.buy.token,
+        };
+
+        *scores.entry(token_pair).or_default() += Score(score);
+    }
+    Ok(scores)
+}
+
+pub struct Config {
+    pub max_winners: usize,
+    pub weth: WrappedNativeToken,
```

Comment:
Do we need to add a[ upper cap](https://github.com/fhenneke/comb_auctions/blob/867e6c318401a1b9b2f90054ccacfce6403abcec/mechanism.py#L435) for rewards?

Reply:
The cap is only needed to compute solver rewards (i.e. how much CoW token they should get based on their reference score). This is luckily not the responsibility of the autopilot.